keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,468 @@
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
+  i.e. scheduler filters Nodes on the domains that have matching Pods in the `Filter` phase.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have matching Pods.
+  i.e. scheduler gives a high score to Nodes on the domains that don't have matching Pods in the `Score` phase.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 10 Nodes.
+
+## Design Details
+
+### API
+
+New parameter called `MinDomains` is introduced.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // MinDomains describes the minimum number of domains.
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have matching Pods.
+  //   i.e. scheduler filters Nodes on the domains that have matching Pods in the `Filter` phase.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have matching Pods.
+  //   i.e. scheduler gives a high score to Nodes on the domains that don't have matching Pods in the `Score` phase.
+  // +optional
+  MinDomains int32
+}
+```
+
+### How user stories are addressed
+
+Users can set `MinDomains` and `whenUnsatisfiable: DoNotSchedule` to achieve it.
+
+```yaml
+spec:
+  topologySpreadConstraints:
+    - minDomains: 10
+      topologyKey: node
+      whenUnsatisfiable: DoNotSchedule
+      labelSelector:
+        matchLabels:
+          foo: bar
+```
+
+With this, until 10 Pods have been created, scheduler schedules each Pod to a different Node.
+This means that, for example, if there are only 5 Nodes that can be scheduled, only a maximum of 5 Pods will be created.
+
+### Test Plan
+
+To ensure this feature to be rolled out in high quality. Following tests are mandatory:
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests / E2E Tests:** All user cases discussed in this KEP must
+  be covered by either integration tests or e2e tests.
+- **Benchmark Tests:** We can bear with slight performance overhead if users are
+  using this feature, but it shouldn't impose penalty to users who are not using
+  this feature. We will verify it by designing some benchmark tests.
+
+### Graduation Criteria
+
+#### Alpha (v1.24):
+
+- [ ] Add new parameter `NinDomains` to `TopologySpreadConstraint` and future gating.
+- [ ] Score extension point implementation.
+- [ ] Filter extension point implementation.
+- [ ] Unit test cases mentioned in the [Test Plan](#test-plan).
+
+#### Beta (v1.25):
+
+- [ ] This feature will be enabled by default as a Beta feature in v1.25.
+- [ ] Add necessary integration/end-to-end tests.
```

Comment:
We usually require integration tests for Alpha. It would be great to have E2E tests that include a cluster autoscaler, but I'm not sure if that's possible.